packages/EasyCore/src/Doctrine/Dispatchers/DeferredEntityEventDispatcher.php

```diff
@@ -39,90 +44,123 @@ public function __construct(EventDispatcherInterface $eventDispatcher)
         }
 
-        /** @var object[] $mergedEntityInsertions */
-        $mergedEntityInsertions = \array_merge(
-            $this->entityInsertions[$transactionNestingLevel] ?? [],
-            $entityInsertions
-        );
-        $this->entityInsertions[$transactionNestingLevel] = $mergedEntityInsertions;
+        $oid = \spl_object_hash($object);
+        $this->entityInsertions[$oid] = $object;
+        $this->entityChangeSets[$transactionNestingLevel][$oid] = $entityChangeSet;
     }
 
-    public function deferUpdates(array $entityUpdates, int $transactionNestingLevel): void
+    /**
+     * @inheritdoc
+     */
+    public function deferUpdate(int $transactionNestingLevel, object $object, array $entityChangeSet): void
     {
         if ($this->enabled === false) {
             return;
         }
 
-        /** @var object[] $mergedEntityUpdates */
-        $mergedEntityUpdates = \array_merge(
-            $this->entityUpdates[$transactionNestingLevel] ?? [],
-            $entityUpdates
-        );
-        $this->entityUpdates[$transactionNestingLevel] = $mergedEntityUpdates;
+        $oid = \spl_object_hash($object);
+        $this->entityUpdates[$oid] = $object;
+        $this->entityChangeSets[$transactionNestingLevel][$oid] = $entityChangeSet;
     }
 
     public function disable(): void
     {
         $this->enabled = false;
+        $this->clear();
     }
 
     public function dispatch(): void
     {
-        $entityInsertions = $this->entityInsertions;
-        $entityUpdates = $this->entityUpdates;
-
-        $this->clear();
-
         if ($this->enabled === false) {
             return;
         }
 
-        $processedEntities = [];
-        foreach ($entityInsertions as $entities) {
-            foreach ($entities as $oid => $entity) {
-                $processedEntities[$oid] = $entity;
-                $this->eventDispatcher->dispatch(new EntityCreatedEvent($entity));
+        try {
+            $mergedEntityChangeSets = [];
+            foreach ($this->entityChangeSets as $levelChangeSets) {
+                foreach ($levelChangeSets as $oid => $changeSet) {
+                    $mergedEntityChangeSets[$oid] = $this->mergeChangeSet(
+                        $mergedEntityChangeSets[$oid] ?? [],
+                        $changeSet
+                    );
+                }
             }
-        }
 
-        foreach ($entityUpdates as $entities) {
-            foreach ($entities as $oid => $entity) {
-                if (isset($processedEntities[$oid])) {
-                    continue;
-                }
-                $processedEntities[$oid] = $entity;
-                $this->eventDispatcher->dispatch(new EntityUpdatedEvent($entity));
+            foreach ($mergedEntityChangeSets as $oid => $entityChangeSet) {
+                $event = $this->createEntityEvent($oid, $entityChangeSet);
+
+                $this->eventDispatcher->dispatch($event);
             }
+        } finally {
+            $this->clear();
```

Comment:
Why does this approach better then the previous one?

Reply:
1. I need `entityInsertions` and `entityUpdates` in a separate method `createEntityEvent`. In the previous approach, these fields are empty.
2. This approach is a bit more optimal as it avoids a temporary copy of fields.